src/fu-engine-repair.c

```diff
@@ -0,0 +1,219 @@
+/*
+ * Copyright (C) 2015 Kate Hsuan <hpa@redhat.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#define G_LOG_DOMAIN "FuEngine"
+
+#include "config.h"
+
+#include <fwupdplugin.h>
+
+#include <fcntl.h>
+#include <glib/gi18n.h>
+#include <glib/gstdio.h>
+#ifdef HAVE_GIO_UNIX
+#include <gio/gunixfdlist.h>
+#include <glib-unix.h>
+#endif
+#include <locale.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+#include "fu-console.h"
+#include "fu-engine.h"
+#include "fu-plugin-private.h"
+#include "fu-engine-repair.h"
+#include "fu-util-common.h"
+
+
+static gboolean
+grubby_set(gboolean enable, const gchar *grubby_arg,  GError **error)
+{
+	g_autofree gchar *output = NULL;
+	g_autofree gchar *arg_string = NULL;
+	const gchar *argv_grubby[] = {"",
+				      "--update-kernel=DEFAULT",
+				      "",
+				      NULL};
+	g_autofree gchar *grubby = NULL;
+	grubby = fu_path_find_program("grubby", NULL);
+	argv_grubby[0] = grubby;
+
+	if (enable)
+		arg_string = g_strdup_printf("--args=%s", grubby_arg);
+	else
+		arg_string = g_strdup_printf("--remove-args=%s", grubby_arg);
+	
+	argv_grubby[2] = arg_string;
+
+	if (!g_spawn_sync(NULL,
+			  (gchar **)argv_grubby,
+			  NULL,
+			  G_SPAWN_DEFAULT,
+			  NULL,
+			  NULL,
+			  &output,
+			  NULL,
+			  NULL,
+			  error))
+		return FALSE;
+
+	return TRUE;
+}
+
+static gboolean 
+grubby_set_lockdown (gboolean enable, GError **error)
+{
+	if (enable)
+		return grubby_set(TRUE, "lockdown=confidentiality", error);
+	else
+		return grubby_set(FALSE, "lockdown=confidentiality", error);
+}
+
+static gboolean
+grubby_set_iomme(gboolean enable, GError **error)
+{
+	if (enable)
+		return grubby_set(TRUE, "iommu=force", error);
+	else
+		return grubby_set(FALSE, "iommu=force", error);
```

Comment:
This needs extra safety checks, what if they have `iommu=soft` for example on kernel command line already?  or `amd_iommu=off` or intel_iommu=sm_off`?  It's a lot of combinations that they can possibly get into invalid states.

Reply:
I could check the security attribute first. If the iommu is invalid then the fix will be applied.

Reply:
For the purpose of "undo" I think you need to detect that the kernel command line has `iommu=force` on it and only undo if that's there.

For detecting whether it's safe/you can do I think you should do:

1. Confirm you have grubby.
2. Check if the kernel command line has any `intel_iommu=`, `amd_iommu=`, or `iommu=` on it already.  If so, it's not safe to apply `iommu=force`.
3. It might make sense to confirm that the system has a `DMAR` or `IVRS` ACPI table (this might need discussion)
4. Check that the kernel has `CONFIG_IOMMU_SUPPORT` enabled (this also might need discussion)

Reply:
Thank you for describing the detail. I'll check the parameters and kernel features before fixing it.